### PR TITLE
INF-506 add cloudfront header

### PR DESCRIPTION
**Pull request by cat5inthecradle**
Prepares us for restricting or identifying traffic to the load balancers. Allows us to identify valid traffic through Cloudfront, vs traffic direct to the load balancer.

Secret generate the secret in all AWS environments (staging, test, levelbuilder, production). I opted NOT to add an `if load_balancer` conditional to the Cloudfront distribution rule. This header will never be used in environments without a load balancer, but I think the consistency will make things simpler in the long run.

## Links

- jira ticket: INF-506

## Testing story

requests from cloudfront to load balancers in staging/test should begin to have these headers. Might be able to see this in the Load Balancer access logs?

## Follow-up work

- Manually test adding a listener rule that will restrict traffic to the load balancers to only requests with the new header.

```erb
  <%w(HTTP HTTPS).each do |protocol|
    CloudfrontOnly<%=protocol%>ListenerRule:
    Type: AWS::ElasticLoadBalancingV2::ListenerRule
    Properties:
      ListenerArn: !Ref <%=protocol%>Listener
      Actions:
        - Type: forward
          TargetGroupArn: !Ref ALBTargetGroup
      Conditions:
        - Field: http-header
          HttpHeaderConfig:
            HttpHeaderName: X-Cdo-Cloudfront-Allow
            Values:
              - '{{resolve:secretsmanager:<%=environment%>/cdo/cloudfront_secret_header_value}}'
    <% end -%>
```

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: lib/cdo/aws/cloudfront.rb [hunk 1]
```diff
@@ -135,6 +135,19 @@ def self.distribution_config(app, origin, aliases, ssl_cert=nil)
       # Add root-domain aliases to production environment stack.
       aliases += CONFIG[app][:aliases] if rack_env?(:production)
 
+      origin_custom_headers = [
+        {
+          HeaderName: 'X-Cdo-INF506',
```
Comment (sureshc) [inline lib/cdo/aws/cloudfront.rb]: Can we choose a header name that explains what its purpose is? `x-Cdo-Auth-CloudFront`? Or `x-Cdo-Auth-Fowarded-By-CloudFront`?

File: lib/cdo/aws/cloudfront.rb [hunk 2]
```diff
@@ -135,6 +135,19 @@ def self.distribution_config(app, origin, aliases, ssl_cert=nil)
...
+          HeaderValue: '17e071a4eeb0'
```
Comment (sureshc) [inline lib/cdo/aws/cloudfront.rb]: I think we can declare a configuration setting in `/config.yml.erb` that is populated from an AWS Secret. Something like
`auth_forwarded_by_cloudfront: !Secret`

And then populate that in the template with something like:
`HeaderValue: '{{resolve:secretsmanager:<%=environment%>/cdo/auth_forwarded_by_cloudfront}}'`

Comment (sureshc) [reply lib/cdo/aws/cloudfront.rb]: We would need to generate the secret values for each environment locally and then publish them to Secrets Manager from a dev environment [via these steps](https://github.com/code-dot-org/code-dot-org/blob/staging-next/config/secrets.md).

File: aws/cloudformation/cloud_formation_stack.yml.erb [hunk 3]
```diff
@@ -168,6 +168,30 @@ Resources:
...
+  Properties:
+    Actions:
+      - Type: forward
+        TargetGroupArn: !Ref ALBTargetGroup
+    Conditions:
+      - Field: http-header
+        HttpHeaderConfig:
+          HttpHeaderName: X-Cdo-Cloudfront-Allow
+          Values:
+            - !Join ['', ['{{resolve:secretsmanager:', !Ref ListenerRuleSecretHeader, ':SecretString:header-value}}' ]]
```
Comment (cat5inthecradle) [inline aws/cloudformation/cloud_formation_stack.yml.erb]: we should deploy without this first.

File: aws/cloudformation/cloud_formation_stack.yml.erb [hunk 4]
```diff
@@ -168,6 +168,30 @@ Resources:
...
+
+  ListenerRuleSecretHeader:
+    Type: 'AWS::SecretsManager::Secret'
+    Properties:
+      Name: ListenerRuleSecretHeader
```
Comment (sureshc) [inline aws/cloudformation/cloud_formation_stack.yml.erb]: Do we need a unique Name for the Secret? Maybe prefix `ListenerRule` with the fully qualified domain name of the CloudFront Distribution?

File: lib/cdo/aws/cloudfront.rb [hunk 5]
```diff
@@ -322,5 +287,72 @@ def self.signed_cookies(resource, expiration_date)
...
+    # `app` is a symbol containing the app name (:pegasus, :dashboard or :hourofcode)
+    # `origin` is the domain name of the backend origin (ex. "origin-autoscale-prod.code.org")
+    def self.get_origins_config(app, origin)
+      raise 'missing CDO.cloudfront_secret_header_value' unless CDO.cloudfront_secret_header_value
+
+      return [
+        *HTTP_CACHE.keys.map do |app_name|
+          # TODO: is `:hourofcode` an option? We're inside a `(Dashboard Pegasus).each do |app|`
+          # block from cloud_formation_stack.yml.erb. Can this code be removed?
+          origin_app = (app == :hourofcode) ? :pegasus : app
```
Comment (cat5inthecradle) [inline lib/cdo/aws/cloudfront.rb:309]: I'm curious if this can be simplified.

**Pull request closed without merging**